### ISSUE TITLE
Report Structure A7

### ISSUE THREAD

**Issue by NF997** [RUG-IDS/Course-2018#58]
The post last week explaining the expected report structure was very helpful.
Would it be possible to publish something similar for this week's report?

**Comment by rm-bwana** [comment]
Hi,

We shall look into it. In the meantime try and adapt the structure from last week to suit this assignment.

Rob.

**Comment by sreejita-rug** [comment]
Hola @NF997 , 
First, sorry for the delay in replying (Temporary lack of internet at home)
So here is the answer in not a nutshell. 

**Introduction**: 
Describe the dataset first (how many features originally, is there any missing data, how many subjects in heathy and how many patients, whether there is a class imbalance or not- all these information is there in the assignment itself). Then describe as 'motivation' that you will investigate this dataset, i) in order to predict the labels of the unlabelled subjects / ii) understand how machine learning can be used in computer aided diagnosis / iii) win a flight ticket to Hawaii / iv) impress the Doctor and hitchhike a ride in the TARDIS / v) educate Trump about AML/ vi) a better reason which we, being the graders, are maybe not sharing with you yet (whichever reason seems good to you).
State that in order to do the prediction one needs to understand which kind(s) of classifier might work well on this dataset and hence state the necessity of descriptive and exploratory analysis of the original dataset. Mention which descriptive exploratory analyses you would be performing. State which kind of scaling will work best on this data log transform / simple data centering/ z-score transform/ anything else), and why. 
**Methodology:** 
1.1 a) State how ANOVA and Eigen value decomposition are different and why both may/may not be necessary in understanding the dataset. With the appropriate visualization tools show in this section itself the ANOVA based findings and the EVD based findings. Is there any elbow in the barplots showing the eigen values? If so where is it approximately? If not (i.e., if all eigen values are of similar magnitude), what does that mean? (Do remember to center the data prior to EVD and ANOVA in order to prevent... remember the stuff in the first 2 lectures? ;)  ). From the findings of this part try to answer the questions asked in **1.1 a**
1.1 b) Briefly differentiate between PCA and tSNE. Describe tSNE briefly (and yes you might have tried n different settings  - maybe n1 different perplexity, n2 different final dimensions, n3 different types of pairwise distances / dissimilarities, but include just the best among all your trials and state the setting of this successful trial). What is particularly helpful is if one brushes through the manual of tSNE made by the guy who made tSNE :)  Wrt the (hyper) parameter perplexity, instead of getting perplexed peruse [this](https://distill.pub/2016/misread-tsne/). From the findings of tSNE try to answer the questions asked in 1.1 b
**1.1 c)** Don't trust our words about the importance of preprocessing. You shouldn't. Not just like that. Atleast not before finding out whether you really need preprocessing. So perform ANOVA, EVD (PCA is based on EVD- _Remember remember the lectures from September!_) once without doing the z-score transformation /log transformation (or however you had scaled your data). Compare the results of the feature selection when done on scaled data and on unscaled data. Similarly compare the effect of scaling of the dataset when performing feature extraction (EVD). From the findings of this part try to answer the questions asked in 1.1 c

**1.2 Experiments**
_Pssst: Eventhough we have mentioned kNN and DT you are welcome to try some other classifiers as well as long as you state why you are planning to try them. Try reading the PLOS paper we uploaded for you all even though you do not want to attempt the BONUS.... might be still helpful... or not. Upto you!_
State what value of **_n_** you have chosen for the n-fold cross validation. Describe if you have chosen any kind of strategy to handle the imbalance - any kind of sampling or anything you took care of while splitting the training data for cross-validation, and why/ why not. 
[**REMEMBER:** In the ith fold of cross validation, while doing z-score transformation take the mean and std of the training set of ith fold keeping the ith test fold data separate (unseen). Use the std and mean from the ith training set to scale the ith test set. The unlabelled set is the generalization set. This should also be scaled in the ith fold using the mean and std of the ith fold's training data. ]
**a) Baseline & more settings:** In the baseline setting use the classifiers mentioned and anything else you want to, on the scaled data without reducing any dimensions. Next do some further preprocessing based on your insights gained from 1.1 and state which are these insights and therefore which changes are you proposing to make on the scale data. 
You may have tried different classifiers and different settings on each classifier (for example, SVM with linear kernel and RBF, kNN with a few different n or a few different metrics, or number of DTs in random forests or level of prunung of DTs) . Tabulate these as 'Experiments done' (1.2 c), in the form of flow chart or hierarchical tree or block diagrams, showing how you preprocessed the data: which scaling, which feature selection/ extraction methods tried, which classifier(s) getting which data- suppose kNN is fed the scaled data with baseline settings once, and scaled data with reduced dimensions once (Fig 1 of assignment can be used as an inspiration). 
Briefly describe the classifiers used and each of the hyperparameters you have used in these classifiers (whether the default or a differently chosen one) and why. (Plots findings will be in Results section)
**b) Ensemble:** Describe the settings used in the built random forest - which features, which hyper-parameters, how many trees, etc
d) _Suggestion for 1.2 d Use mode ('mode' as in the mode from basic Statistics) on the predictions from your top 5 or 7 best classifiers (one of those odd cases where odd is better than even, but this is just a friendly suggestion) to get 'The One Final List To Define Them All'- list of labels for your generalization set (this is what you will submit as Team_x_prediction.csv separately (1.2 d))._
**Results**
Using appropriate evaluation methods^ - accuracy or precision or area under the curve or a combination of these, tabulate the 2 or 3 best results from each of the classifiers you tried (assuming that for each classifier you have tweaked at least 1 hyperparameter to improve the classifier's prediction). If there is space and you have the wish and time to, you can tabulate everything and just highlight the 2 or 3 best ones, or if you want you can put the results from the not-so -efficient classifiers in the appendix. 
^ Use 1 or 2 simple sentences to briefly describe each of the evaluation measures you have used.

You can also use ROC curves to compare the classifier performances - but bear in mind that the default colours of the plotting functions are sometimes confusing - often the best and worst curves are assigned the same (or similar) colours by the sadist MATLAB/ Python/R and thus we are left to wonder which is what and contemplate the question to 42. So take control dear students! If not yet of your entire lives, the Universe and everything, atleast of the colour schemes in the plots for IDS assignment report. You can do it :D This is your plot, your report, your hard work, why let Matlab/ Python or R's 'default'settings confuse you and then confuse us, eh! You define!
So the above tables and plots will give you an idea of the best classifiers. 
Use these best classifiers to get predicted labels of the unlabelled (generalization) set. Use the same preprocessing done to the training data fed to these selected classifiers.
d) Do submit your 'One Final List To Define Them All' (Team_x_prediction.csv) as a separate .csv file and do NOT share the list in the report itself. This will lead to unnecessary wastage of space and hinder fast checking of labels by us

**Discussion:** 
1.2 a) Do you observe overfitting in the baseline? How would you recognize it and how will you prevent it in each of the classifier for which it showed up? Discuss whether further preprocessing of the dataset improved the classifier performance from that of the baseline settings. Refer to the plots and finding in the Results section for this. Which settings (classifier + hyperparameter + scaling and/or  feature selection / extraction method and/or imbalance handling strategy) were best and which were worst (based on their auc/ precision-recall/ sensitivity/ specificity values? Which of these evaluation methods would you trust more in case there is a discrepancy among them, and why? What do you infer from all these? 

**Conclusion:** 
Mention your take-home-message from all your investigations on this dataset, whether something else can be further investigated, which kind of classifiers would you advise for such kind of data and if you were to name 1 (or 2) classifier and setting explicitly as the best one, which one(s) would that(/they) be and why?

[If we do not see any plots or tabulated results which you refer to, to discuss your findings we can't grade you for that question. The results serve as proof/evidence you use to fight your case with... I mean to substantiate your discussion and conclusion with ]
